packages/suite/src/actions/suite/metadataActions.ts

```diff
@@ -672,6 +718,259 @@ export const addMetadata = (payload: MetadataAddPayload) => (dispatch: Dispatch)
+
+    if (!provider) {
+        return;
+    }
+
+    const createBatchPromises = missingFiles.map(async ({ fileName, aesKey }) => {
+        const encrypted = await metadataUtils.encrypt(
+            {
+                version: '1.0.0',
+                dummyField: Date.now(),
+            },
+            aesKey,
+        );
+
+        return {
+            fileName: `${fileName}_${METADATA_ENCODING_VERSION}`,
+            content: encrypted,
+        };
+    });
+
+    const batchToUpload = await Promise.all(createBatchPromises);
+
+    try {
+        const response = await provider.batchSetFileContent(batchToUpload);
+
+        if (!response.success) {
+            throw new Error(response.error);
+        }
+    } catch (err) {
+        const error = provider.error('OTHER_ERROR', err.message);
+        return dispatch(handleProviderError(error, ProviderErrorAction.SAVE));
+    }
+};
+
+type MigratedDeviceFile = {
+    newFileName: EnabledDeviceMetadata['fileName'];
+    newAesKey: EnabledDeviceMetadata['aesKey'];
+    data: Pick<EnabledDeviceMetadata, 'walletLabel'>;
+};
+type MigratedAccountFile = {
+    newFileName: AccountMetadata['fileName'];
+    newAesKey: AccountMetadata['aesKey'];
+    data: Pick<AccountMetadata, 'accountLabel' | 'addressLabels' | 'outputLabels'>;
+};
+type MigratedFile = MigratedDeviceFile | MigratedAccountFile;
+
+const migrateMetadataFiles =
+    (missingFiles: Array<MissingFileInfo>) => async (dispatch: Dispatch, getState: GetState) => {
+        // get old keys
+        await dispatch(setDeviceMetadataKey(1, true));
+        // did user confirm labeling on device? or maybe device was not connected
+        // so suite does not have keys and needs to stop here
+        if (getState().suite.device?.metadata.status !== 'enabled') {
+            throw new Error('Labeling rejected by the user');
+        }
+        // if yes, add metadata keys to accounts
+        dispatch(syncMetadataKeys());
+
+        // try fetching data for entities with missing files
+        const filesToMigrate: Array<MigratedFile> = [];
+
+        const { device } = getState().suite;
+        if (device?.metadata.status !== 'enabled') {
+            return;
+        }
+
+        const provider = dispatch(getProvider());
+        if (!provider) {
+            throw new Error('No provider found');
+        }
+
+        const missingDeviceFile = missingFiles.find(
+            ({ type }) => type === 'device',
+        ) as MissingDeviceFile;
+
+        if (missingDeviceFile) {
+            const providerResponse = await provider.getFileContent(device?.metadata.fileName);
+
+            if (providerResponse.success && providerResponse.payload) {
+                try {
+                    const data = metadataUtils.decrypt(
+                        metadataUtils.arrayBufferToBuffer(providerResponse.payload),
+                        device?.metadata.aesKey,
+                    );
+
+                    if (data?.walletLabel) {
```

Comment:
why `if (data?.walletLabel) {` ? doesn't it make sense to simply migrate every file?

Reply:
I was trying to be more thorough but I suppose it doesn't bring much value, true.

Reply:
done